docs/custom-filters.adoc

```diff
@@ -45,23 +151,163 @@ Instances are specific to the channel between a single client and a single broke
+        mutateResponse(response, context);
+        context.forwardResponse(header, response);
+    }
+}
+----
+
+==== Responding from a Request Filter without Proxying
+
+In some cases we may wish to not forward a message from the client to Kafka. Instead, we want to generate a response
+message in Kroxylicious and send it back to the client. For example:
+
+[source,java]
+----
+import org.apache.kafka.common.message.ApiVersionsRequestData;
+import org.apache.kafka.common.message.ApiVersionsResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.protocol.Errors;
+
+public class ApiVersionsErrorFilter implements ApiVersionsRequestFilter {
+
+    @Override
+    public void onApiVersionsRequest(short apiVersion, RequestHeaderData header, ApiVersionsRequestData request, KrpcFilterContext context) {
+        ApiVersionsResponseData response = new ApiVersionsResponseData();
+        response.setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code());
+        context.forwardResponse(response);
+    }
+}
+----
+This will respond to all Api Versions requests with an error response without forwarding any of those requests to Kafka.
+
+Again, `forwardResponse` must be called synchronously, within the `onXyzRequest` method.
+
+==== Sending new messages to Kafka that are not visible to the Client
+
+Filters can also send arbitrary new messages to Kafka that will be invisible to the client. For example we could sent a
+request to Kafka for topic metadata while handling a FetchRequest. The Filter can then handle the MetadataResponse without
+any other Filters in the chain interfering with the response or it being exposed to the client.
+
+For example:
+
+[source,java]
+----
+public class FetchFilter implements FetchRequestFilter{
+    public static final short METADATA_VERSION_SUPPORTING_TOPIC_IDS = (short) 12;
+
+    @Override
+    public void onFetchRequest(short apiVersion, RequestHeaderData header, FetchRequestData request, KrpcFilterContext context) {
+        MetadataRequestData metadataRequest = new MetadataRequestData();
+        metadataRequest.topics().add(new MetadataRequestData.MetadataRequestTopic().setTopicId(Uuid.randomUuid()));
+        CompletionStage<MetadataResponseData> stage = context.sendRequest(METADATA_VERSION_SUPPORTING_TOPIC_IDS, request);
+        stage.thenAccept(response -> doSomethingWithMetadata(response, METADATA_VERSION_SUPPORTING_TOPIC_IDS));
+        context.forwardRequest(header, request);
+    }
+}
+----
+
+As you can see, we need to know the API version we want our message to be encoded at. This must be done carefully as we
+currently do not have a mechanism for your Filter to know the supported API versions of the upstream Kafka cluster. If
+you use hardcoded apiVersions as shown, your Filter will only work with upstream brokers that support that apiVersion for
+that RPC.
```

Comment:
Maybe we should have a discussion thread or issue started for this and link to that here.

Reply:
done